### ISSUE TITLE
"/" for subfolders should work for substring, not just prefix...

### ISSUE THREAD

**Issue by fredstluka** [RealRaven2000/QuickFolders#135]
When I use Shift-M, Shift-T or Shift-J to go to a subfolder of a specified folder, I should be allowed to use a substring, not just a prefix.

I hit Shift-M, Shift-T or Shift-J to see a textbox.
I type "snipp/".
I hope to see the subfolder of my "01_Snippets" folder.
But it says: "No matching parent folders!"
Instead I have to type "01_snipp"

It would be nice to be able to type the substring "snipp" instead of the prefix "01_snipp".

Thanks!
--Fred

**Comment by RealRaven2000** [comment]
I did that deliberately because you can end up with a load of (or even some) confusing suggestions otherwise, so it's actually easier for most users - it makes sense to use this for the child / target folder. But I understand that you chose numeric prefixes and then it's obviously not so great in that case. I think the nicest thing would be to ignore anything NN_ and treat the string afterwards as the start of the string...

**Comment by fredstluka** [comment]
Axel,

Yeah, ignoring NN_ would solve my problem.  Good idea!

Or even match on the prefix of any the segments of the name
that are separated by underscores.  So, "abc" would match any
of:
- abcde
- abcde_foo_bar
- foo_abcd_bar
- foo_bar_abc

--Fred
------------------------------------------------------------------------
Fred Stluka -- http://bristle.com -- Glad to be of service!
Open Source: Without walls and fences, we need no Windows or Gates.
------------------------------------------------------------------------

**Comment by RealRaven2000** [comment]
> Axel, Yeah, ignoring NN_ would solve my problem.  Good idea! Or even match on the prefix of any the segments of the name that are separated by underscores.  So, "abc" would match any of: - abcde - abcde_foo_bar - foo_abcd_bar - foo_bar_abc
> […](#)

yes - I like that idea the most, I think this would also technically be the "easiest" to implement (not super trivial, but doable). I think it should do both space and "_".

Do you think it should also work with "-" (dash) for things like composite names such as "Armin Mueller-Stahl"? So it would be found by entering "actors/st". The main reason I narrow down the results is that using the format "`start of parent/start of child folder`" is very intuitive and avoids too many false positives

**Comment by fredstluka** [comment]
Axel,

Yeah, I was going to suggest also "-" (hyphen), and maybe also
"." (period).  But I didn't know offhand if they were allowed in
folder names.

--Fred
------------------------------------------------------------------------
Fred Stluka -- http://bristle.com -- Glad to be of service!
Open Source: Without walls and fences, we need no Windows or Gates.
------------------------------------------------------------------------

**Comment by RealRaven2000** [comment]
Just testing. I made a subfolder Add-ons/01_xyz-test and searched for ad/xy:

![image](https://user-images.githubusercontent.com/6314674/114440412-01238d80-9bc2-11eb-8d6c-be96e17cf2fb.png)

..and it actually comes up! Relevant code is here in QickFolders.Interfaace.addMatchingFolder(..)

```
if (searchString.length<=2 && matchPos!=0) { // doesn't start with single/two letters?
  // is it the start of a new word? e.g. searching 'F' should match "x-fred" "x fred" "x.fred" "x,fred" ":fred" "(fred" "@fred"
  if (" .-,_:@([".indexOf(folderNameSearched.substr(matchPos-1,1))<0)
    continue;  // skip if not starting with single letter
}
```
So (at least for the search folderParent/someText it should already work. I am actually allowing the delimiters` . , - _ :` and `@`. So what's different in your case? Are you entering the first 2 letters of the parent folder?

**Comment by RealRaven2000** [comment]
Reread the original bug description. I think you want the same match algorithm for all parent folders. That's currently not supported - so I need to roll out the same matching method to the parents. Hope it's not going to completely slow down the search with this additional complexity...

**Comment by RealRaven2000** [comment]
I wrote some test code, but it turns out not very useful - too many results that I don't want to see:

![image](https://user-images.githubusercontent.com/6314674/114446383-151ebd80-9bc9-11eb-8d35-101e27746578.png)

This is especially jarring for folders that are based on domain names using . as delimiters such as sumofus.org. Any comments how I could improve this? I am using a regular expression for the matching, so I don't really want to make the list of accepted delimiters a user-configurable list / parameter. The only thing I can think of is to take "." out of the equation. Test version for Thunderbird 68 (my current production version with 300+ folders so best for testing search performance in a real world enivronment) below:

[QuickFolders-wx-4.21pre19.zip](https://github.com/RealRaven2000/QuickFolders/files/6299147/QuickFolders-wx-4.21pre19.zip)

**Comment by RealRaven2000** [comment]
If I take out matching the delimiters `-` and `.` for parent folders then I get more useful results (and it covers your case). Some search examples:

![image](https://user-images.githubusercontent.com/6314674/114447914-d558d580-9bca-11eb-9287-7ce975307608.png)

![image](https://user-images.githubusercontent.com/6314674/114448088-0507dd80-9bcb-11eb-87e7-f2be02f330b2.png)


I particularly like the fact that I was using `_` to prepend to some folder names to force them showing up on top, and now they are included.

**Comment by RealRaven2000** [comment]
Test version for Thunderbird 78 and later:

[QuickFolders-wx-5.5pre2.zip](https://github.com/RealRaven2000/QuickFolders/files/6303263/QuickFolders-wx-5.5pre2.zip)

**Comment by RealRaven2000** [comment]
@fredstluka did you get a chance to test the above version?

**Comment by fredstluka** [comment]
Axel,

Omitting "." from the set of delimiters is fine with me.  I always
use "_" ( a leftover habit from year of VAX/VMS and DOS/Windows
usage where only one "." was allowed in a filename.  And from
years of programming in various languages where "." and "-"
were not allowed in identifiers.  "_" has always been the delimiter
that worked everywhere.

Yeah, a few years back, I was prefixing things with "_" to move
them to the top.  Then I started needing control over the exact
order of the multiple things I'd moved to the top, so I switched
to 1_, 2_, etc.  Then I needed more than 10 of them, so I
switched to 01_, 02_, etc.

I never got around to testing QuickFolders-wx-5.5pre2.zip 
<https://github.com/RealRaven2000/QuickFolders/files/6303263/QuickFolders-wx-5.5pre2.zip>, 
but
today I tested QuickFolders-wx-5.5pre3.zip 
<https://github.com/RealRaven2000/QuickFolders/files/6303326/QuickFolders-wx-5.5pre3.zip> 
that you sent for
#136 and it adds this feature also.  Perfect!  You can close this
ticket.

Thanks!
--Fred
------------------------------------------------------------------------
Fred Stluka -- http://bristle.com -- Glad to be of service!
Open Source: Without walls and fences, we need no Windows or Gates.
------------------------------------------------------------------------

**Comment by RealRaven2000** [comment]
Implemented  in in 5.5  published today
